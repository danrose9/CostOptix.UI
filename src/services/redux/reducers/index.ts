export { default as serviceProviderReducer } from './serviceProviderSlice';
export { default as resourceReducer } from './resourceSlice';
export { default as costDashboardReducer } from './costDashboardSlice';
export { default as costContainerReducer } from './costContainerSlice';
export { default as applicationReducer } from './applicationSlice';
export { default as userProfileReducer } from './userProfileSlice';
export { default as azureManagementReducer } from './azureManagementSlice';
