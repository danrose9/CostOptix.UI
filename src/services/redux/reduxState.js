export const reduxState = {
  USER_PROFILE: 'userProfile',
  AZURE_MANAGEMENT: 'azureManagement',
  SERVICE_PROVIDERS: 'serviceProviders',
  BILLING_ACCOUNTS: 'billingAccounts',
  COST_DASHBOARD: 'costDashboard',
  RESOURCES: 'resources',
  APPLICATION: 'application',
};
