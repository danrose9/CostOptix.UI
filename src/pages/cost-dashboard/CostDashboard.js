import React, { useState, useContext } from 'react';

import { useSelector } from 'react-redux';
import { reduxState } from '../../services/redux/reduxState';

import { PageTitle } from '../../components/PageTitle';
import { Grid, Divider } from 'semantic-ui-react';
import * as Widget from './index';
import { PageContent } from '../__styles__/DefaultPageStyles';

import { ApplicationContext } from '../../app/ApplicationContext';
import getSymbolFromCurrency from 'currency-symbol-map';

export const CostDashboard = () => {
  const application = useContext(ApplicationContext);

  const billingAccountCurrency = useSelector(
    (state) => state[reduxState.COST_DASHBOARD].currency
  );

  const [isCurrencyConflict, setIsCurrencyConflict] = useState(false);
  const [currency, setCurrency] = useState(null);
  const [currencySymbol, setCurrencySymbol] = useState(null);

  const isCurrencyConflictCallback = (isConflict) => {
    const currencySymbol = () => {
      setIsCurrencyConflict(isConflict);
      if (isConflict) {
        setCurrency(application.settings.preferredCurrency);
      } else {
        setCurrency(billingAccountCurrency);
      }
      return getSymbolFromCurrency(currency);
    };
    setCurrencySymbol(currencySymbol);
  };

  return (
    <PageContent>
      <PageTitle title="Cost Dashboard" />
      <Divider />
      <Grid>
        <Grid.Row>
          <Grid.Column width={12}>
            <Widget.CostDashboardChart currency={currencySymbol} />
          </Grid.Column>
          <Grid.Column width={4}>
            <Widget.CostDasboardEstimates
              isCurrencyConflict={isCurrencyConflict}
              currency={currencySymbol}
            />
          </Grid.Column>
        </Grid.Row>

        <Grid.Row columns="equal">
          <Grid.Column>
            <Widget.MostExpensiveInstance
              isCurrencyConflict={isCurrencyConflict}
            />
          </Grid.Column>
          <Grid.Column>
            <Widget.FastestGrowingInstance />
          </Grid.Column>
          <Grid.Column width={6}>
            <Widget.BillingAccounts
              isCurrencyConflictCallback={isCurrencyConflictCallback}
            />
          </Grid.Column>
        </Grid.Row>
      </Grid>
    </PageContent>
  );
};

export default CostDashboard;
