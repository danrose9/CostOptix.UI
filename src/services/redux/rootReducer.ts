import { combineReducers } from 'redux';
import { reduxState } from './reduxState';

import userProfileReducer from './reducers/userProfileSlice';
import azureManagementReducer from './reducers/azureManagementSlice';
import serviceProviderReducer from './reducers/serviceProvidersSlice';
import costDashboardReducer from './reducers/costDashboardSlice';
import resourceReducer from './reducers/resourceSlice';
import applicationReducer from './reducers/applicationSlice';

const appReducer = combineReducers({
  [reduxState.USER_PROFILE]: userProfileReducer,
  [reduxState.AZURE_MANAGEMENT]: azureManagementReducer,
  [reduxState.SERVICE_PROVIDERS]: serviceProviderReducer,
  [reduxState.COST_DASHBOARD]: costDashboardReducer,
  [reduxState.RESOURCES]: resourceReducer,
  [reduxState.APPLICATION]: applicationReducer,
});

export const rootReducer = (state: any, action: any) => {
  if (action.type === 'USER_LOGOUT') {
    state = undefined;
  }

  return appReducer(state, action);
};

export type IRootState = ReturnType<typeof rootReducer>;
