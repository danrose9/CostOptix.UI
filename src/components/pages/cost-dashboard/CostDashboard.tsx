import React, { useState, useContext } from 'react';
import { useSelector } from 'react-redux';
import { reduxState } from '../../../services/redux/reduxState';
import PageWrapper from '../PageWrapper';
import { Grid, Divider } from 'semantic-ui-react';
import * as Widget from './index';
import { ApplicationContext } from '../../../app/ApplicationContext';
import getSymbolFromCurrency from 'currency-symbol-map';
import { IRootState } from '../../../services/redux/rootReducer';

export const CostDashboard = () => {
  const application = useContext(ApplicationContext);

  const billingAccountCurrency = useSelector((state: IRootState) => state[reduxState.COST_DASHBOARD].currency);

  const [isCurrencyConflict, setIsCurrencyConflict] = useState<boolean>(false);
  const [currency, setCurrency] = useState<string>('');
  const [currencySymbol, setCurrencySymbol] = useState<string | undefined>('');

  const isCurrencyConflictCallback = (isConflict: boolean) => {
    const currencySymbol = () => {
      setIsCurrencyConflict(isConflict);
      if (isConflict) {
        setCurrency(application.settings.preferredCurrency);
      } else {
        setCurrency(billingAccountCurrency);
      }
      return getSymbolFromCurrency(currency);
    };
    setCurrencySymbol(currencySymbol);
  };

  return (
    <>
      <PageWrapper title="Cost Dashboard">
        <Divider />
        <Grid>
          <Grid.Row>
            <Grid.Column width={12} product-tour="tour-start">
              <Widget.CostDashboardChart currency={currencySymbol} />
            </Grid.Column>
            <Grid.Column width={4}>
              <Widget.CostDasboardEstimates isCurrencyConflict={isCurrencyConflict} currency={currencySymbol} />
            </Grid.Column>
          </Grid.Row>
          <Grid.Row columns="equal">
            <Grid.Column product-tour="most-expensive">
              <Widget.MostExpensiveInstance isCurrencyConflict={isCurrencyConflict} />
            </Grid.Column>
            <Grid.Column>
              <Widget.FastestGrowingInstance />
            </Grid.Column>
            <Grid.Column width={5} product-tour="active-providers">
              <Widget.BillingAccounts isCurrencyConflictCallback={isCurrencyConflictCallback} />
            </Grid.Column>
          </Grid.Row>
        </Grid>
      </PageWrapper>
    </>
  );
};

export default CostDashboard;
