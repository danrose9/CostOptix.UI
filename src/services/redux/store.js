import { configureStore } from '@reduxjs/toolkit';
import storage from 'redux-persist/lib/storage/session';
import {
  persistStore,
  persistReducer,
  FLUSH,
  REHYDRATE,
  PAUSE,
  PERSIST,
  PURGE,
  REGISTER,
} from 'redux-persist';
import { rootReducer } from './rootReducer';
import monitorReducersEnhancer from './enhancers/monitorReducer';
import loggerMiddleware from './enhancers/logger';

const persistConfig = {
  key: 'root',
  version: 1,
  storage,
};

const persistedReducer = persistReducer(persistConfig, rootReducer);

const store = configureStore({
  reducer: persistedReducer,
  middleware: (getDefaultMiddleware) =>
    getDefaultMiddleware({
      serializableCheck: {
        ignoredActions: [
          FLUSH,
          REHYDRATE,
          PAUSE,
          PERSIST,
          PURGE,
          REGISTER,
          'Profile/incrementLoginCount/fulfilled',
        ],
      },
    }).concat(loggerMiddleware),
  enhancers: [monitorReducersEnhancer],
});

let persistor = persistStore(store);

export { store, persistor };

// export type AppDispatch = typeof store.dispatch
