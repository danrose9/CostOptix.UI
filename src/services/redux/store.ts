import { configureStore } from '@reduxjs/toolkit';
import storage from 'redux-persist/lib/storage/session';
import { persistStore, persistReducer, FLUSH, REHYDRATE, PAUSE, PERSIST, PURGE, REGISTER } from 'redux-persist';
import { rootReducer } from './rootReducer';
import monitorReducersEnhancer from './enhancers/monitorReducer';
import loggerMiddleware from './enhancers/logger';
import { useDispatch } from 'react-redux';

const persistConfig = {
  key: 'root',
  version: 1,
  storage,
};

const persistedReducer = persistReducer(persistConfig, rootReducer);

const store = configureStore({
  reducer: persistedReducer,
  middleware: (getDefaultMiddleware) =>
    getDefaultMiddleware({
      serializableCheck: {
        ignoredActions: [FLUSH, REHYDRATE, PAUSE, PERSIST, PURGE, REGISTER, 'Profile/incrementLoginCount/fulfilled'],
      },
    }).concat(loggerMiddleware),
  enhancers: [monitorReducersEnhancer],
  devTools: process.env.NODE_ENV === 'development',
});

let persistor = persistStore(store);

// Infer the `RootState` and `AppDispatch` types from the store itself
export type RootState = ReturnType<typeof store.getState>;

// Inferred type: {posts: PostsState, comments: CommentsState, users: UsersState}
export type AppDispatch = typeof store.dispatch;
export const useAppDispatch: () => AppDispatch = useDispatch;

export { store, persistor };
