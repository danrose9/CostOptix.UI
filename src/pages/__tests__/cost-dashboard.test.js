import React from 'react';

import { render, cleanup } from '@testing-library/react';
import CostDashboard from '../cost-dashboard/CostDashboard';
import {
  MostExpensiveInstance,
  FastestGrowingInstance,
} from '../cost-dashboard';
import renderer from 'react-test-renderer';
import { BrowserRouter as Router } from 'react-router-dom';

import { Provider } from 'react-redux';
import { store } from '../../services/redux/store';

afterEach(() => {
  cleanup();
});

describe('Cost Dashboard', () => {
  test('should render Cost Dashboard', () => {
    render(
      <Provider store={store}>
        <Router>
          <CostDashboard />
        </Router>
      </Provider>
    );
  });

  test('matches snapshot', () => {
    const tree = renderer
      .create(
        <Provider store={store}>
          <Router>
            <CostDashboard />
          </Router>
        </Provider>
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });
});

describe('Most Expensive resource widget', () => {
  test('should render widget', () => {
    render(
      <Provider store={store}>
        <Router>
          <MostExpensiveInstance isCurrencyConflict={false} />
        </Router>
      </Provider>
    );
  });

  test('column width is correct', () => {
    const { getByTestId } = render(
      <Provider store={store}>
        <Router>
          <MostExpensiveInstance isCurrencyConflict={true} />
        </Router>
      </Provider>
    );

    expect(getByTestId('header-cell-1')).toHaveProperty('colSpan', 4);
  });
});

describe('Fastest Growing resource widget', () => {
  test('should render widget', () => {
    render(
      <Provider store={store}>
        <Router>
          <FastestGrowingInstance />
        </Router>
      </Provider>
    );
  });

  // test click header
  // test click row
});
