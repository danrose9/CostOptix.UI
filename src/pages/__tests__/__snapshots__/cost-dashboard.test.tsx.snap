// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Cost Dashboard matches snapshot 1`] = `
<div
  className="sc-bcXHqe jMdjRw"
>
  <div
    className="sc-dkrFOg feMyyx"
  >
    <div
      className="sc-jSUZER dYOwPx"
    >
      Cost Dashboard
    </div>
  </div>
  <div
    className="ui divider"
  />
  <div
    className="ui grid"
  >
    <div
      className="row"
    >
      <div
        className="twelve wide column"
      >
        <div
          className="ui orange loading segment"
        >
          <div
            className="recharts-responsive-container"
            style={
              Object {
                "height": 350,
                "maxHeight": undefined,
                "minHeight": undefined,
                "minWidth": undefined,
                "width": "100%",
              }
            }
          />
        </div>
      </div>
      <div
        className="four wide column"
      >
        <div
          className="ui olive loading segment"
        >
          <table
            className="ui selectable table"
          >
            <thead
              className=""
            >
              <tr
                className=""
              >
                <th
                  className=""
                  colSpan="3"
                >
                  Month-to-Date Cost Estimates
                </th>
              </tr>
            </thead>
            <tbody
              className=""
            />
            <tfoot
              className="full-width"
            >
              <tr
                className=""
              >
                <td
                  className=""
                >
                  Total
                </td>
                <td
                  className=""
                />
                <td
                  className="right aligned"
                >
                  0
                </td>
              </tr>
            </tfoot>
          </table>
        </div>
      </div>
    </div>
    <div
      className="equal width row"
    >
      <div
        className="column"
      >
        <div
          className="ui orange loading segment"
        >
          <table
            className="ui fixed selectable table"
          >
            <thead
              className=""
              onClick={[Function]}
            >
              <tr
                className=""
              >
                <th
                  className="sc-csuSiG lgsuTZ"
                  colSpan={3}
                  data-testid="header-cell-1"
                >
                  Most Expensive Resources Last 30 days
                </th>
              </tr>
            </thead>
            <thead
              className=""
            >
              <tr
                className=""
              >
                <th
                  className=""
                >
                  Resource
                </th>
                <th
                  className=""
                >
                  Service
                </th>
                <th
                  className="right aligned"
                >
                  Cost
                </th>
              </tr>
            </thead>
            <tbody
              className=""
            />
          </table>
        </div>
      </div>
      <div
        className="column"
      >
        <div
          className="ui purple loading segment"
        >
          <table
            className="ui fixed selectable table"
          >
            <thead
              className=""
              data-testid="table-header-1"
              onClick={[Function]}
            >
              <tr
                className=""
              >
                <th
                  className="sc-csuSiG lgsuTZ"
                  colSpan="3"
                >
                  Fastest Growing Resources
                </th>
              </tr>
            </thead>
            <thead
              className=""
            >
              <tr
                className=""
              >
                <th
                  className=""
                >
                  Resource
                </th>
                <th
                  className=""
                >
                  Service
                </th>
                <th
                  className="right aligned"
                >
                  30d growth
                </th>
              </tr>
            </thead>
            <tbody
              className=""
            />
          </table>
        </div>
      </div>
      <div
        className="five wide column"
      >
        <div
          className="ui green segment"
        >
          <table
            className="ui selectable table"
          >
            <thead
              className="full-width"
            >
              <tr
                className=""
              >
                <th
                  className=""
                  colSpan="3"
                >
                  Active Billing Accounts
                </th>
                <th
                  className="center aligned two wide"
                >
                  <i
                    aria-hidden="true"
                    className="refresh disabled fitted loading icon sc-kDvujY izENdF"
                    onClick={[Function]}
                  />
                </th>
              </tr>
            </thead>
            <tbody
              className=""
              style={
                Object {
                  "height": "30px",
                }
              }
            />
            <tfoot
              className="full-width"
            >
              <tr
                className=""
              >
                <th
                  className=""
                  colSpan="4"
                >
                  Last Updated: 
                </th>
              </tr>
            </tfoot>
          </table>
        </div>
      </div>
    </div>
  </div>
</div>
`;
